akka-cluster/src/main/scala/akka/cluster/ClusterHeartbeat.scala

```diff
@@ -168,6 +177,261 @@ private[cluster] final class ClusterHeartbeatSender extends Actor with ActorLogg
+  val heartbeatTask = scheduler.schedule(
+    PeriodicTasksInitialDelay max HeartbeatInterval,
+    HeartbeatInterval, self, HeartbeatTick)
+
+  override def preStart(): Unit = {
+    cluster.subscribe(self, classOf[MemberEvent], classOf[ReachabilityEvent])
+  }
+
+  override def postStop(): Unit = {
+    state.activeReceivers.foreach(a ⇒ crossDcFailureDetector.remove(a.address))
+    heartbeatTask.cancel()
+    cluster.unsubscribe(self)
+  }
+
+  /**
+   * Looks up and returns the remote cluster heartbeat connection for the specific address.
+   */
+  def heartbeatReceiver(address: Address): ActorSelection =
+    context.actorSelection(RootActorPath(address) / "system" / "cluster" / "heartbeatReceiver")
+
+  def receive: Actor.Receive = {
+    log.info(s"$incLocalClockJson Initialized as dormant in ${selfDatacenter}")
+    dormant orElse introspecting
+  }
+
+  /**
+   * In this state no cross-datacenter heartbeats are sent by this actor.
+   * This may be because one of those reasons:
+   *   - no nodes in other DCs were detected yet
+   *   - nodes in other DCs are present, but this node is not tht n-th oldest in this DC (see
+   *     `number-of-cross-datacenter-monitoring-actors`), so it does not have to monitor that other data centers
+   *
+   * In this state it will however listen to cluster events to eventually take over monitoring other DCs
+   * in case it becomes "old enough".
+   */
+  def dormant: Actor.Receive = {
+    case s: CurrentClusterState ⇒ init(s)
+    case MemberRemoved(m, _)    ⇒ removeMember(m)
+    case evt: MemberEvent       ⇒ addMember(evt.member)
+    case HeartbeatTick          ⇒ // ignore...
+  }
+
+  def active: Actor.Receive = {
+    case HeartbeatTick                ⇒ heartbeat()
+    case HeartbeatRsp(from)           ⇒ heartbeatRsp(from)
+    case MemberRemoved(m, _)          ⇒ removeMember(m)
+    case evt: MemberEvent             ⇒ addMember(evt.member)
+    case UnreachableMember(m)         ⇒ unreachableMember(m)
+    case ReachableMember(m)           ⇒ reachableMember(m)
+    case ExpectedFirstHeartbeat(from) ⇒ triggerFirstHeartbeat(from)
+  }
+
+  def introspecting: Actor.Receive = {
+    case ReportStatusTo(o) ⇒
+      log.debug(s"$incLocalClockJson Updated status observer to {} (activelyMonitoring:{})", o, activelyMonitoring)
+      if (activelyMonitoring) o.get ! CrossDcHeartbeatSender.MonitoringActive(state)
+      else o.get ! CrossDcHeartbeatSender.MonitoringDormant()
+      this.observer = o
+  }
+
+  def init(snapshot: CurrentClusterState): Unit = {
+    val nodes = snapshot.members.filter(filterExternalMembers).map(_.uniqueAddress)
+    val unreachable = snapshot.unreachable.filter(filterExternalMembers).map(_.uniqueAddress)
+    state = state.init(nodes, unreachable)
+  }
+
+  def addMember(m: Member): Unit = {
+    val memberIsCandidateForMonitoring =
+      m.status == MemberStatus.Up && // only Up members are considered for addition, since we require the oldest relationships to hold 
```

Comment:
I think we must use those that are leaving and such until they are removed, so it's more like
```
m.status != MemberStatus.Joining && m.status != MemberStatus.WeaklyUp
```